File: tmdbService.ts
/**
 * ═══════════════════════════════════════════════════════
 * TMDB API INTEGRATED SERVICE - MiM Plus
 * ═══════════════════════════════════════════════════════
 * 
 * خدمة متكاملة للاتصال بواجهة برمجة تطبيقات TMDB والحصول على البيانات الحية.
 * مصممة للعمل في بيئة JavaScript / TypeScript (Project IDX) للهواتف الذكية.
 * خالية تماماً من البيانات الوهمية (No Mock Data)، وتعتمد بالكامل على طلبات الشبكة الحقيقية.
 * 
 * المميزات:
 * 1. استخراج الحقول المطلوبة بدقة عالية وبوستر عالي الجودة.
 * 2. جلب التفاصيل الكاملة للأعمال (الوصف، المواسم، الحلقات، الممثلين، المخرجين وصورهم).
 * 3. نظام صفحات (Pagination) ديناميكي لـ fetchTrendingContent.
 * 4. دعم كامل للغة العربية والتعريب التلقائي لأسماء التصنيفات.
 * 5. معالجة أخطاء try...catch احترافية ومفصلة.
 */

// ═══════════════════════════════════════════════════════
// القائمة المرجعية لتعريفات التصنيفات باللغة العربية
// ═══════════════════════════════════════════════════════
const GENRE_MAP: Record<number, string> = {
  28: "أكشن",
  12: "مغامرة",
  16: "رسوم متحركة",
  35: "كوميديا",
  80: "جريمة",
  99: "وثائقي",
  18: "دراما",
  10751: "عائلي",
  14: "فانتازيا",
  36: "تاريخ",
  27: "رعب",
  10402: "موسيقى",
  9648: "غموض",
  10749: "رومانسي",
  878: "خيال علمي",
  10770: "تلفزيوني",
  53: "إثارة",
  10752: "حرب",
  37: "غرب أمريكي",
  10759: "أكشن ومغامرة",
  10762: "أطفال",
  10763: "أخبار",
  10764: "واقعي",
  10765: "خيال وفانتازيا",
  10766: "أوبرا صابونية",
  10767: "برنامج حواري",
  10768: "حرب وسياسة"
};

// ═══════════════════════════════════════════════════════
// تَعْرِيف واجهات البيانات (Interfaces)
// ═══════════════════════════════════════════════════════

export interface TrendingItem {
  id: number;
  title: string;
  posterUrl: string;
  rating: number;
  category: string;
  releaseYear: number;
  mediaType: "movie" | "tv";
}

export interface CastMember {
  id: number;
  name: string;
  role: string; // Actor (Character name) or Director
  profileUrl: string | null;
}

export interface WorkDetails {
  id: number;
  title: string;
  overview: string;
  posterUrl: string;
  rating: number;
  releaseYear: number;
  category: string;
  seasonsCount: number | null;  // خاص بالمسلسلات (null للأفلام)
  episodesCount: number | null; // خاص بالمسلسلات (null للأفلام)
  cast: CastMember[];           // الممثلون
  directors: CastMember[];      // المخرجون
}

// ═══════════════════════════════════════════════════════
// فِئَة الخدمة الرَّئِيسِيَّة (TMDB Service Class)
// ═══════════════════════════════════════════════════════

export class TmdbService {
  private baseUrl: string = "https://api.themoviedb.org/3";
  private apiKey: string;
  private imageBaseUrl: string = "https://image.tmdb.org/t/p";

  /**
   * إنشاء كائن الخدمة.
   * @param apiKey مفتاح TMDB API. يمكنك الحصول عليه مجاناً من إعدادات حسابك على TMDB.
   */
  constructor(apiKey: string) {
    if (!apiKey || apiKey.trim() === "") {
      throw new Error("عذراً، يجب تقديم مفتاح API Key صالح لـ TMDB للاتصال بالخدمة الحية.");
    }
    this.apiKey = apiKey;
  }

  /**
   * دالة مساعدة لترميز وضمان روابط الصور عالية الجودة
   */
  private getPosterUrl(path: string | null, size: "w500" | "h632" | "original" = "w500"): string {
    return path ? `${this.imageBaseUrl}/${size}${path}` : "https://placehold.co/500x750/091e1a/ffffff?text=No+Poster";
  }

  /**
   * دالة مساعدة لترميز صور طاقم العمل
   */
  private getProfileUrl(path: string | null): string | null {
    return path ? `${this.imageBaseUrl}/w185${path}` : null;
  }

  /**
   * جلب المحتوى الرائج (مسلسلات وأفلام) مع التصفح والصفحات الديناميكية.
   * 
   * @param mediaType نوع المحتوى المطلق ('all' للاثنين معاً، أو 'movie' أو 'tv')
   * @param page رقم الصفحة لتفعيل الـ Pagination الديناميكي
   * @returns قائمة بالكائنات المهيأة للواجهة والمستخرجة بدقة.
   */
  async fetchTrendingContent(
    mediaType: "all" | "movie" | "tv" = "all",
    page: number = 1
  ): Promise<TrendingItem[]> {
    try {
      const url = `${this.baseUrl}/trending/${mediaType}/week?api_key=${this.apiKey}&page=${page}&language=ar`;
      const response = await fetch(url);

      if (!response.ok) {
        throw new Error(`فشل الاتصال بـ TMDB: الكود ${response.status} (${response.statusText})`);
      }

      const data = await response.json();
      
      if (!data.results || !Array.isArray(data.results)) {
        return [];
      }

      return data.results.map((item: any) => {
        // تحديد العنوان الأساسي بناءً على نوع العمل (أفلام أو مسلسلات)
        const title = item.title || item.name || item.original_title || item.original_name || "عنوان غير معروف";
        
        // جلب السنة بدقة
        const releaseDate = item.release_date || item.first_air_date;
        const releaseYear = releaseDate ? new Date(releaseDate).getFullYear() : 0;

        // ترجمة التصنيفات
        const firstGenreId = item.genre_ids && item.genre_ids[0];
        const category = firstGenreId ? (GENRE_MAP[firstGenreId] || "منوع") : "منوع";

        return {
          id: item.id,
          title: title,
          posterUrl: this.getPosterUrl(item.poster_path, "w500"),
          rating: Number((item.vote_average || 0).toFixed(1)),
          category: category,
          releaseYear: releaseYear,
          mediaType: item.media_type || (mediaType === "all" ? "movie" : mediaType)
        };
      });

    } catch (error: any) {
      console.error(`[Error in fetchTrendingContent]: ${error.message || error}`);
      throw error; // إعادة توجيه الخطأ لمعالجته في الواجهة الرسومية
    }
  }

  /**
   * جلب تفاصيل العمل الفني بشكل كامل وحقيقي مع طاقم العمل.
   * 
   * @param mediaType نوع العمل ('movie' للفيلم، أو 'tv' للمسلسل)
   * @param id رمز التعريف الخاص بالعمل (ID)
   * @returns تفاصيل العمل الكاملة مهيأة للنسخ والاستخدام الفوري.
   */
  async fetchDetails(mediaType: "movie" | "tv", id: number): Promise<WorkDetails> {
    try {
      // جلب تفاصيل العمل الأساسية مع طاقم العمل في طلب شبكة واحد عبر append_to_response لسرعة فائقة وتحسين الأداء
      const detailsUrl = `${this.baseUrl}/${mediaType}/${id}?api_key=${this.apiKey}&append_to_response=credits&language=ar`;
      const response = await fetch(detailsUrl);

      if (!response.ok) {
        throw new Error(`فشل جلب تفاصيل المعرف ${id} من TMDB: كود ${response.status}`);
      }

      const data = await response.json();

      const title = data.title || data.name || data.original_title || data.original_name || "عمل فني";
      const releaseDate = data.release_date || data.first_air_date;
      const releaseYear = releaseDate ? new Date(releaseDate).getFullYear() : 0;
      
      // جلب التصنيف
      const category = data.genres && data.genres.length > 0 ? data.genres[0].name : "منوع";

      // تصفية فريق التمثيل والتمثيليات (المخرجين والممثلين)
      const cast: CastMember[] = [];
      const directors: CastMember[] = [];

      if (data.credits) {
        // 1. الممثلين الرائدين (أول 8 ممثلين فقط من أجل المظهر الجمالي والأداء)
        if (Array.isArray(data.credits.cast)) {
          data.credits.cast.slice(0, 8).forEach((actor: any) => {
            cast.push({
              id: actor.id,
              name: actor.name,
              role: actor.character || "ممثل",
              profileUrl: this.getProfileUrl(actor.profile_path)
            });
          });
        }

        // 2. المخرجين (طاقم الإخراج)
        if (Array.isArray(data.credits.crew)) {
          data.credits.crew.forEach((crewMember: any) => {
            if (crewMember.job === "Director" || crewMember.job === "Executive Producer" || crewMember.department === "Directing") {
              directors.push({
                id: crewMember.id,
                name: crewMember.name,
                role: crewMember.job || "مخرج",
                profileUrl: this.getProfileUrl(crewMember.profile_path)
              });
            }
          });
        }
      }

      return {
        id: data.id,
        title: title,
        overview: data.overview || "لا يوجد وصف متوفر باللغة العربية لهذا العمل حالياً.",
        posterUrl: this.getPosterUrl(data.poster_path, "original"), // دقة أصلية لتفاصيل العمل
        rating: Number((data.vote_average || 0).toFixed(1)),
        releaseYear: releaseYear,
        category: category,
        seasonsCount: mediaType === "tv" ? (data.number_of_seasons || 1) : null,
        episodesCount: mediaType === "tv" ? (data.number_of_episodes || 12) : null,
        cast: cast,
        directors: directors.slice(0, 3) // أول 3 مخرجين بارزين كحد أقصى لتناسق الواجهة
      };

    } catch (error: any) {
      console.error(`[Error in fetchDetails for ID ${id}]: ${error.message || error}`);
      throw error;
    }
  }
}

File: app/src/main/java/com/example/TmdbService.kt
package com.example

import android.util.Log
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okhttp3.Request
import org.json.JSONObject
import java.io.IOException

// ═══════════════════════════════════════════════════════
// DATA CLASSES MATCHING THE TS SERVICE
// ═══════════════════════════════════════════════════════
data class TmdbTrendingItem(
    val id: Int,
    val title: String,
    val posterUrl: String,
    val rating: Double,
    val category: String,
    val releaseYear: Int,
    val mediaType: String // "movie" or "tv"
)

data class TmdbCastMember(
    val id: Int,
    val name: String,
    val role: String,
    val profileUrl: String?
)

data class TmdbWorkDetails(
    val id: Int,
    val title: String,
    val overview: String,
    val posterUrl: String,
    val rating: Double,
    val releaseYear: Int,
    val category: String,
    val seasonsCount: Int?, // null for movie
    val episodesCount: Int?, // null for movie
    val cast: List<TmdbCastMember>,
    val directors: List<TmdbCastMember>
)

class TmdbService(private val apiKey: String) {
    private val client = OkHttpClient()
    private val baseUrl = "https://api.themoviedb.org/3"
    private val imageBaseUrl = "https://image.tmdb.org/t/p"

    private val genreMap = mapOf(
        28 to "أكشن",
        12 to "مغامرة",
        16 to "رسوم متحركة",
        35 to "كوميديا",
        80 to "جريمة",
        99 to "وثائقي",
        18 to "دراما",
        10751 to "عائلي",
        14 to "فانتازيا",
        36 to "تاريخ",
        27 to "رعب",
        10402 to "موسيقى",
        9648 to "غموض",
        10749 to "رومانسي",
        878 to "خيال علمي",
        10770 to "تلفزيوني",
        53 to "إثارة",
        10752 to "حرب",
        37 to "غرب أمريكي",
        10759 to "أكشن ومغامرة",
        10762 to "أطفال",
        10763 to "أخبار",
        10764 to "واقعي",
        10765 to "خيال وفانتازيا",
        10766 to "أوبرا صابونية",
        10767 to "برنامج حواري",
        10768 to "حرب وسياسة"
    )

    private fun getPosterUrl(path: String?, size: String = "w500"): String {
        return if (!path.isNullOrEmpty() && path != "null") {
            "$imageBaseUrl/$size$path"
        } else {
            "https://placehold.co/500x750/091e1a/ffffff?text=No+Poster"
        }
    }

    private fun getProfileUrl(path: String?): String? {
        return if (!path.isNullOrEmpty() && path != "null") {
            "$imageBaseUrl/w185$path"
        } else {
            null
        }
    }

    suspend fun fetchTrendingContent(
        mediaType: String = "all", // "all", "movie", "tv"
        page: Int = 1
    ): List<TmdbTrendingItem> = withContext(Dispatchers.IO) {
        try {
            val url = "$baseUrl/trending/$mediaType/week?api_key=$apiKey&page=$page&language=ar"
            val request = Request.Builder().url(url).build()
            client.newCall(request).execute().use { response ->
                if (!response.isSuccessful) throw IOException("Failed to load TMDB trending: raw code ${response.code}")
                val bodyString = response.body?.string() ?: return@use emptyList()
                val json = JSONObject(bodyString)
                val results = json.optJSONArray("results") ?: return@use emptyList()
                val list = mutableListOf<TmdbTrendingItem>()
                for (i in 0 until results.length()) {
                    val item = results.getJSONObject(i)
                    val id = item.optInt("id")
                    
                    val title = item.optString("title").takeIf { it.isNotEmpty() && it != "null" }
                        ?: item.optString("name").takeIf { it.isNotEmpty() && it != "null" }
                        ?: item.optString("original_title").takeIf { it.isNotEmpty() && it != "null" }
                        ?: item.optString("original_name").takeIf { it.isNotEmpty() && it != "null" }
                        ?: "عنوان غير معروف"
                        
                    val releaseDate = item.optString("release_date").takeIf { it.isNotEmpty() && it != "null" }
                        ?: item.optString("first_air_date").takeIf { it.isNotEmpty() && it != "null" }
                        ?: ""
                    val releaseYear = if (releaseDate.length >= 4) {
                        try { releaseDate.substring(0, 4).toInt() } catch (e: Exception) { 0 }
                    } else {
                        0
                    }

                    val genreIds = item.optJSONArray("genre_ids")
                    val firstGenreId = if (genreIds != null && genreIds.length() > 0) genreIds.getInt(0) else -1
                    val category = genreMap[firstGenreId] ?: "منوع"

                    val posterPath = item.optString("poster_path").takeIf { it.isNotEmpty() && it != "null" }
                    val rating = item.optDouble("vote_average", 0.0)
                    val itemMediaType = item.optString("media_type", if (mediaType == "all") "movie" else mediaType)

                    list.add(
                        TmdbTrendingItem(
                            id = id,
                            title = title,
                            posterUrl = getPosterUrl(posterPath, "w500"),
                            rating = String.format("%.1f", rating).toDoubleOrNull() ?: 0.0,
                            category = category,
                            releaseYear = releaseYear,
                            mediaType = itemMediaType
                        )
                    )
                }
                return@use list
            }
        } catch (e: Exception) {
            Log.e("TmdbService", "Error in fetchTrendingContent", e)
            throw e
        }
    }

    suspend fun searchContent(
        query: String,
        page: Int = 1
    ): List<TmdbTrendingItem> = withContext(Dispatchers.IO) {
        try {
            val encodedQuery = java.net.URLEncoder.encode(query, "UTF-8")
            val url = "$baseUrl/search/multi?api_key=$apiKey&query=$encodedQuery&page=$page&language=ar"
            val request = Request.Builder().url(url).build()
            client.newCall(request).execute().use { response ->
                if (!response.isSuccessful) throw IOException("Failed to search TMDB: raw code ${response.code}")
                val bodyString = response.body?.string() ?: return@use emptyList()
                val json = JSONObject(bodyString)
                val results = json.optJSONArray("results") ?: return@use emptyList()
                val list = mutableListOf<TmdbTrendingItem>()
                for (i in 0 until results.length()) {
                    val item = results.getJSONObject(i)
                    val id = item.optInt("id")
                    
                    val itemMediaType = item.optString("media_type", "movie")
                    if (itemMediaType != "movie" && itemMediaType != "tv") continue // Skip people
                    
                    val title = item.optString("title").takeIf { it.isNotEmpty() && it != "null" }
                        ?: item.optString("name").takeIf { it.isNotEmpty() && it != "null" }
                        ?: item.optString("original_title").takeIf { it.isNotEmpty() && it != "null" }
                        ?: item.optString("original_name").takeIf { it.isNotEmpty() && it != "null" }
                        ?: "عنوان غير معروف"
                        
                    val releaseDate = item.optString("release_date").takeIf { it.isNotEmpty() && it != "null" }
                        ?: item.optString("first_air_date").takeIf { it.isNotEmpty() && it != "null" }
                        ?: ""
                    val releaseYear = if (releaseDate.length >= 4) {
                        try { releaseDate.substring(0, 4).toInt() } catch (e: Exception) { 0 }
                    } else {
                        0
                    }

                    val genreIds = item.optJSONArray("genre_ids")
                    val firstGenreId = if (genreIds != null && genreIds.length() > 0) genreIds.getInt(0) else -1
                    val category = genreMap[firstGenreId] ?: "منوع"

                    val posterPath = item.optString("poster_path").takeIf { it.isNotEmpty() && it != "null" }
                    val rating = item.optDouble("vote_average", 0.0)

                    list.add(
                        TmdbTrendingItem(
                            id = id,
                            title = title,
                            posterUrl = getPosterUrl(posterPath, "w500"),
                            rating = String.format("%.1f", rating).toDoubleOrNull() ?: 0.0,
                            category = category,
                            releaseYear = releaseYear,
                            mediaType = itemMediaType
                        )
                    )
                }
                return@use list
            }
        } catch (e: Exception) {
            Log.e("TmdbService", "Error in searchContent", e)
            throw e
        }
    }

    suspend fun fetchDetails(
        mediaType: String, // "movie" or "tv"
        id: Int
    ): TmdbWorkDetails = withContext(Dispatchers.IO) {
        try {
            val url = "$baseUrl/$mediaType/$id?api_key=$apiKey&append_to_response=credits&language=ar"
            val request = Request.Builder().url(url).build()
            client.newCall(request).execute().use { response ->
                if (!response.isSuccessful) throw IOException("Failed to load details for $id: raw code ${response.code}")
                val bodyString = response.body?.string() ?: throw IOException("Empty body")
                val data = JSONObject(bodyString)

                val title = data.optString("title").takeIf { it.isNotEmpty() && it != "null" }
                    ?: data.optString("name").takeIf { it.isNotEmpty() && it != "null" }
                    ?: data.optString("original_title").takeIf { it.isNotEmpty() && it != "null" }
                    ?: data.optString("original_name").takeIf { it.isNotEmpty() && it != "null" }
                    ?: "عمل فني"

                val releaseDate = data.optString("release_date").takeIf { it.isNotEmpty() && it != "null" }
                    ?: data.optString("first_air_date").takeIf { it.isNotEmpty() && it != "null" }
                    ?: ""
                val releaseYear = if (releaseDate.length >= 4) {
                    try { releaseDate.substring(0, 4).toInt() } catch (e: Exception) { 0 }
                } else {
                    0
                }

                // Get category from genres array
                val genresArr = data.optJSONArray("genres")
                val category = if (genresArr != null && genresArr.length() > 0) {
                    genresArr.getJSONObject(0).optString("name", "منوع")
                } else {
                    "منوع"
                }

                val castList = mutableListOf<TmdbCastMember>()
                val directList = mutableListOf<TmdbCastMember>()

                val credits = data.optJSONObject("credits")
                if (credits != null) {
                    val castArr = credits.optJSONArray("cast")
                    if (castArr != null) {
                        val limit = minOf(castArr.length(), 10)
                        for (i in 0 until limit) {
                            val actor = castArr.getJSONObject(i)
                            castList.add(
                                TmdbCastMember(
                                    id = actor.optInt("id"),
                                    name = actor.optString("name", "غير معروف"),
                                    role = actor.optString("character", "ممثل"),
                                    profileUrl = getProfileUrl(actor.optString("profile_path").takeIf { it.isNotEmpty() && it != "null" })
                                )
                            )
                        }
                    }

                    val crewArr = credits.optJSONArray("crew")
                    if (crewArr != null) {
                        for (i in 0 until crewArr.length()) {
                            val crew = crewArr.getJSONObject(i)
                            val job = crew.optString("job")
                            val dept = crew.optString("department")
                            if (job == "Director" || job == "Executive Producer" || dept == "Directing") {
                                directList.add(
                                    TmdbCastMember(
                                        id = crew.optInt("id"),
                                        name = crew.optString("name", "غير معروف"),
                                        role = job.takeIf { it.isNotEmpty() } ?: "مخرج",
                                        profileUrl = getProfileUrl(crew.optString("profile_path").takeIf { it.isNotEmpty() && it != "null" })
                                    )
                                )
                            }
                        }
                    }
                }

                val seasonsCount = if (mediaType == "tv") data.optInt("number_of_seasons", 1) else null
                val episodesCount = if (mediaType == "tv") data.optInt("number_of_episodes", 12) else null

                val posterPath = data.optString("poster_path").takeIf { it.isNotEmpty() && it != "null" }
                val rating = data.optDouble("vote_average", 0.0)

                return@use TmdbWorkDetails(
                    id = data.optInt("id"),
                    title = title,
                    overview = data.optString("overview", "لا يوجد وصف متوفر باللغة العربية لهذا العمل حالياً."),
                    posterUrl = getPosterUrl(posterPath, "original"),
                    rating = String.format("%.1f", rating).toDoubleOrNull() ?: 0.0,
                    releaseYear = releaseYear,
                    category = category,
                    seasonsCount = seasonsCount,
                    episodesCount = episodesCount,
                    cast = castList,
                    directors = directList.take(3)
                )
            }
        } catch (e: Exception) {
            Log.e("TmdbService", "Error in fetchDetails", e)
            throw e
        }
    }
}
